File: src/test/java/im/tox/tox4j/core/LoadSaveTest.java
package im.tox.tox4j.core;

import im.tox.tox4j.ToxCoreImplTestBase;
import im.tox.tox4j.ToxCoreTestBase$;
import im.tox.tox4j.core.enums.ToxStatus;
import im.tox.tox4j.exceptions.ToxException;
import org.junit.Test;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import static org.junit.Assert.*;

public final class LoadSaveTest extends ToxCoreImplTestBase {

    private interface Check {
        boolean change(ToxCore tox) throws ToxException;
        void check(ToxCore tox);
    }

    private void testLoadSave(Check check) throws Exception {
        boolean moreTests = true;
        while (moreTests) {
            byte[] data;
            try (ToxCore tox = newTox()) {
                moreTests = check.change(tox);
                data = tox.save();
            }
            try (ToxCore tox = newTox(data)) {
                check.check(tox);
            }
        }
    }

    @Test
    public void testName() throws Exception {
        testLoadSave(new Check() {
            private byte[] expected = {};

            @Override
            public boolean change(ToxCore tox) throws ToxException {
                if (expected == null) {
                    expected = new byte[0];
                } else {
                    expected = ToxCoreTestBase$.MODULE$.randomBytes(expected.length + 1);
                }
                tox.setName(expected);
                return expected.length < ToxConstants.MAX_NAME_LENGTH;
            }

            @Override
            public void check(ToxCore tox) {
                assertArrayEquals(expected, tox.getName());
            }
        });
    }

    @Test
    public void testStatusMessage() throws Exception {
        testLoadSave(new Check() {
            private byte[] expected = null;

            @Override
            public boolean change(ToxCore tox) throws ToxException {
                if (expected == null) {
                    expected = new byte[0];
                } else {
                    expected = ToxCoreTestBase$.MODULE$.randomBytes(expected.length + 1);
                }
                tox.setStatusMessage(expected);
                return expected.length < ToxConstants.MAX_NAME_LENGTH;
            }

            @Override
            public void check(ToxCore tox) {
                assertArrayEquals(expected, tox.getStatusMessage());
            }
        });
    }

    @Test
    public void testStatus() throws Exception {
        testLoadSave(new Check() {
            private final List<ToxStatus> expected = new ArrayList<>(Arrays.asList(ToxStatus.values()));

            @Override
            public boolean change(ToxCore tox) throws ToxException {
                tox.setStatus(expected.get(expected.size() - 1));
                return expected.size() > 1;
            }

            @Override
            public void check(ToxCore tox) {
                assertEquals(expected.remove(expected.size() - 1), tox.getStatus());
            }
        });
    }

    @Test
    public void testNoSpam() throws Exception {
        testLoadSave(new Check() {
            private int expected = -1;

            @Override
            public boolean change(ToxCore tox) throws ToxException {
                tox.setNospam(++expected);
                return expected < 100;
            }

            @Override
            public void check(ToxCore tox) {
                assertEquals(expected, tox.getNospam());
            }
        });
    }

    @Test
    public void testFriend() throws Exception {
        testLoadSave(new Check() {
            private int expected;

            @Override
            public boolean change(ToxCore tox) throws ToxException {
                try (ToxCore toxFriend = newTox()) {
                    expected = tox.addFriend(toxFriend.getAddress(), "hello".getBytes());
                }
                return false;
            }

            @Override
            public void check(ToxCore tox) {
                assertEquals(1, tox.getFriendList().length);
                assertEquals(expected, tox.getFriendList()[0]);
            }
        });
    }

    @Test
    public void testSaveNotEmpty() throws Exception {
        try (ToxCore tox = newTox()) {
            byte[] data = tox.save();
            assertNotNull(data);
            assertNotEquals(0, data.length);
        }
    }

    @Test
    public void testSaveRepeatable() throws Exception {
        try (ToxCore tox = newTox()) {
            assertArrayEquals(tox.save(), tox.save());
        }
    }

    @Test
    public void testLoadSave1() throws Exception {
        byte[] data = newTox().save();
        assertArrayEquals(newTox(data).save(), newTox(data).save());
    }

    @Test
    public void testLoadSave2() throws Exception {
        byte[] data = newTox().save();
        assertEquals(data.length, newTox(data).save().length);
    }

    @Test
    public void testLoadSave3() throws Exception {
        byte[] data = newTox().save();
        assertArrayEquals(data, newTox(data).save());
    }

}

File: src/test/java/im/tox/tox4j/core/LoadSaveTest.scala
package im.tox.tox4j.core

import java.util.{ ArrayList, Arrays }

import im.tox.tox4j.ToxCoreTestBase
import im.tox.tox4j.core.ToxCoreFactory.withTox
import im.tox.tox4j.core.enums.ToxUserStatus
import im.tox.tox4j.core.options.SaveDataOptions
import im.tox.tox4j.exceptions.ToxException
import org.junit.Assert._
import org.junit.Test

final class LoadSaveTest extends ToxCoreTestBase {

  private trait Check {
    @throws(classOf[ToxException[_]])
    def change(tox: ToxCore): Boolean
    def check(tox: ToxCore): Unit
  }

  private def testLoadSave(check: Check): Unit = {
    var moreTests = true
    var data: Array[Byte] = new Array[Byte](0)
    while (moreTests) {
      withTox { tox =>
        moreTests = check.change(tox)
        data = tox.getSaveData
      }

      withTox(SaveDataOptions.ToxSave(data)) { tox =>
        check.check(tox)
      }
    }
  }

  @Test def testName(): Unit = {
    testLoadSave(new Check() {
      private var expected: Array[Byte] = null

      @throws(classOf[ToxException[_]])
      override def change(tox: ToxCore): Boolean = {
        if (expected == null) {
          expected = Array[Byte]()
        } else {
          expected = ToxCoreTestBase.randomBytes(expected.length + 1)
        }
        tox.setName(expected)
        expected.length < ToxCoreConstants.MAX_NAME_LENGTH
      }

      override def check(tox: ToxCore): Unit = {
        assertArrayEquals(expected, tox.getName)
      }
    })
  }

  @Test def testStatusMessage(): Unit = {
    testLoadSave(new Check() {
      private var expected: Array[Byte] = null

      @throws(classOf[ToxException[_]])
      override def change(tox: ToxCore): Boolean = {
        if (expected == null) {
          expected = Array[Byte]()
        } else {
          expected = ToxCoreTestBase.randomBytes(expected.length + 1)
        }
        tox.setStatusMessage(expected)
        expected.length < ToxCoreConstants.MAX_NAME_LENGTH
      }

      override def check(tox: ToxCore): Unit = {
        assertArrayEquals(expected, tox.getStatusMessage)
      }
    })
  }

  @Test def testStatus(): Unit = {
    testLoadSave(new Check() {

      private val expected = new ArrayList(Arrays.asList(ToxUserStatus.values(): _*))

      @throws(classOf[ToxException[_]])
      override def change(tox: ToxCore): Boolean = {
        tox.setStatus(expected.get(expected.size() - 1))
        expected.size() > 1
      }

      override def check(tox: ToxCore): Unit = {
        assertEquals(expected.remove(expected.size() - 1), tox.getStatus)
      }
    })
  }

  @Test def testNoSpam(): Unit = {
    testLoadSave(new Check() {
      private var expected = -1

      @throws(classOf[ToxException[_]])
      override def change(tox: ToxCore): Boolean = {
        expected += 1
        tox.setNoSpam(expected)
        expected < 100
      }

      override def check(tox: ToxCore): Unit = {
        assertEquals(expected, tox.getNoSpam)
      }
    })
  }

  @Test def testFriend(): Unit = {
    testLoadSave(new Check() {
      private var expected: Int = 1

      @throws(classOf[ToxException[_]])
      override def change(tox: ToxCore): Boolean = {
        withTox { toxFriend =>
          expected = tox.addFriend(toxFriend.getAddress, "hello".getBytes)
        }
        false
      }

      override def check(tox: ToxCore): Unit = {
        assertEquals(1, tox.getFriendList.length)
        assertEquals(expected, tox.getFriendList(0))
      }
    })
  }

  @Test def testSaveNotEmpty(): Unit = {
    withTox { tox =>
      var data = tox.getSaveData
      assertNotNull(data)
      assertNotEquals(0, data.length)
    }
  }

  @Test def testSaveRepeatable(): Unit = {
    withTox { tox =>
      assertArrayEquals(tox.getSaveData, tox.getSaveData)
    }
  }

  @Test def testLoadSave1(): Unit = {
    withTox { tox =>
      var data = tox.getSaveData
      var data1: Array[Byte] = new Array[Byte](0)
      var data2: Array[Byte] = new Array[Byte](0)
      withTox(SaveDataOptions.ToxSave(data)) { tox1 =>
        data1 = tox1.getSaveData
      }
      withTox(SaveDataOptions.ToxSave(data)) { tox2 =>
        data2 = tox2.getSaveData
      }
      assertArrayEquals(data1, data2)
    }
  }

  @Test def testLoadSave2(): Unit = {
    withTox { tox =>
      var data = tox.getSaveData
      withTox(SaveDataOptions.ToxSave(data)) { tox1 =>
        assertEquals(data.length, tox1.getSaveData.length)
      }
    }
  }

  @Test def testLoadSave3(): Unit = {
    withTox { tox =>
      var data = tox.getSaveData
      withTox(SaveDataOptions.ToxSave(data)) { tox1 =>
        assertArrayEquals(data, tox1.getSaveData)
      }
    }
  }

}
